test/test_web.py

```diff
@@ -0,0 +1,69 @@
+# coding=utf-8
+"""Tests Sopel's web tools"""
+from __future__ import unicode_literals, absolute_import, print_function, division
+
+
+from sopel.web import search_urls
+
+
+def test_search_urls():
+    urls = list(search_urls('http://example.com'))
+    assert len(urls) == 1, 'Must find 1 URL, found %d' % len(urls)
+    assert 'http://example.com' in urls
+
+
+def test_search_urls_with_text():
+    urls = list(search_urls('before http://example.com after'))
+    assert len(urls) == 1, 'Must find 1 URL, found %d' % len(urls)
+    assert 'http://example.com' in urls
+
+
+def test_search_urls_multiple_urls():
+    urls = list(search_urls('http://a.com/ http://b.com/'))
+    assert len(urls) == 2, 'Must find 2 URLs, found %d' % len(urls)
+    assert 'http://a.com/' in urls
+    assert 'http://b.com/' in urls
+
+
+def test_search_urls_multiple_urls_unique():
+    urls = list(search_urls('http://a.com/ http://b.com/ http://a.com/'))
+    assert len(urls) == 2, 'Must find 2 URLs, found %d' % len(urls)
+    assert 'http://a.com/' in urls
+    assert 'http://b.com/' in urls
+
+
+def test_search_urls_multiple_urls_unique_keep_ordering():
+    urls = list(
+        search_urls('http://a.com/ http://c.com/ http://b.com/ http://a.com/'))
+    assert len(urls) == 3, 'Must find 3 URLs, found %d' % len(urls)
+    assert 'http://a.com/' in urls
+    assert 'http://b.com/' in urls
+    assert 'http://c.com/' in urls
+    assert urls == [
+        'http://a.com/',
+        'http://c.com/',
+        'http://b.com/',
+    ]
+
+
+def test_search_urls_exclusion_char():
+    urls = list(search_urls('!http://example.com', exclusion_char='!'))
+    assert not urls, 'Must not find URL, found %d' % len(urls)
+
+    urls = list(
+        search_urls(
+            '!http://example.com http://example.com',
+            exclusion_char='!'
+        )
+    )
+    assert len(urls) == 1, 'Must find 1 URL, found %d' % len(urls)
```

Comment:
This isn't a useful test of the exclusion character, since even if the `exclusion_char` doesn't work properly the `search_urls` function will only return one item anyway (because it only returns unique URLs).

Better to double down on the following test, and assert that `http://a.com` is not in the returned list.

Reply:
I agree that it's a very edge-case, but here is my thought process.

First, this is what we do: ignore **this particular instance** of a URL:

    >>> search_urls('!http://a.com http://b.com http://a.com', '!')
    ['http://b.com', 'http://a.com']

Which is fine by me. But it could mean something else: blacklist **all instance** of a URL:

    >>> search_urls('!http://a.com http://b.com http://a.com', '!')
    ['http://b.com']

I think I should change my test to add this extra `http://b.com` URL, maybe that would speak to you?

Reply:
No, it should _not_ blacklist **all instances** of that URL in the same line, haha.

I think I get it—this test is checking that the `exclusion_char` only applies to the URL it abuts, and not all identical URLs in the line. The previous and next tests check that the `exclusion_char` actually excludes the single URL it abuts.

You needn't change the tests. Just a couple comments about what each is testing (since they're grouped under the same function) would be enough.

Reply:
Will do! :)

Reply:
Done, have a look at the new tests! :)